File: leetcode/2. problem/122.买卖股票的最佳时机-ii.java
/**
 * @lc app=leetcode.cn id=122 lang=java
 *
 *     [122] 买卖股票的最佳时机 II
 */

// @lc code=start
class Solution {
    public int maxProfit(int[] prices) {
        // 贪心
        // return extracted(prices);

        // 动态规划
        return extracted2(prices);
    }

    /**
     * Your runtime beats 23.78 % of java submissions
     * Your memory usage beats 8.89 % of java submissions (41.5 MB)
     * 
     * @param prices
     * @return
     */
    private int extracted2(int[] prices) {
        int n = prices.length;
        // 创建二维数组存储状态
        int[][] dp = new int[n][2];
        // 初始状态
        dp[0][0] = 0;
        dp[0][1] = -prices[0];
        for (int i = 1; i < n; ++i) {
            // 第 i 天，没有股票
            dp[i][0] = Math.max(dp[i - 1][0], dp[i - 1][1] + prices[i]);
            // 第 i 天，持有股票
            dp[i][1] = Math.max(dp[i - 1][1], dp[i - 1][0] - prices[i]);
        }
        return dp[n - 1][0]; // 卖出股票收益高于持有股票收益，因此取[0]
    }

    private int extracted(int[] prices) {
        int sum = 0;
        for (int i = 1; i < prices.length; i++) {
            if (prices[i] > prices[i - 1]) {
                sum = sum + (prices[i] - prices[i - 1]);
            }
        }
        return sum;
    }
}
// @lc code=end

File: leetcode/2. problem/122.买卖股票的最佳时机-ii.go
/*
 * @lc app=leetcode.cn id=122 lang=golang
 *
 * [122] 买卖股票的最佳时机 II
 */
package main

// @lc code=start
func maxProfit(prices []int) int {
	// dp[i][0] 代表不持有股票收益
	// dp[i][1] 代表持有股票收益
	return newFunction9(prices)
}

// 200/200 cases passed (0 ms)
// Your runtime beats 100 % of golang submissions
// Your memory usage beats 10.97 % of golang submissions (4 MB)
func newFunction9(prices []int) int {
	length := len(prices)
	dp := make([][]int, length)
	for i := 0; i < length; i++ {
		dp[i] = make([]int, 2)
	}
	var max = func(i, j int) int {
		if i > j {
			return i
		}
		return j
	}

	dp[0][0] = 0
	dp[0][1] = -prices[0]
	for i := 1; i < length; i++ {
		dp[i][0] = max(dp[i-1][0], dp[i-1][1]+prices[i])
		dp[i][1] = max(dp[i-1][1], dp[i-1][0]-prices[i])

	}
	return dp[length-1][0]
}

// @lc code=end
